File: ext/hrr_rb_mount/hrr_rb_mount.c
#include "hrr_rb_mount.h"
#include "sys/mount.h"

VALUE rb_mHrrRbMount;
VALUE rb_mHrrRbMountConst;

/*
 * A wrapper around mount system call.
 *
 * == Synopsis:
 *   HrrRbMount.mount "source", "target", "filesystemtype", mountflags, "data" # => 0
 *
 * @overload mount(source, target, filesystemtype, mountflags=0, data="") => 0
 *   @param source [String] The pathname referring to a device or a pathname of a directory or file, or a dummy string.
 *   @param target [String] The location (a directory or file) specified by the pathname.
 *   @param filesystemtype [String] The filesystem type, which is supported by the kernel.
 *   @param mountflags [Integer] The mount operation is performed depending on the bits specified in the mountflags.
 *   @param data [String] Per filesystem options.
 *   @return [Integer] 0.
 *   @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
 */
VALUE
hrr_rb_mount_mount(int argc, VALUE *argv, VALUE self)
{
  const char *source, *target, *filesystemtype, *data;
  unsigned long mountflags;

  rb_check_arity(argc, 3, 5);

  source         = StringValueCStr(argv[0]);
  target         = StringValueCStr(argv[1]);
  filesystemtype = StringValueCStr(argv[2]);
  mountflags     = argc < 4 ? 0 : NUM2ULONG(argv[3]);
  data           = argc < 5 ? "" : StringValueCStr(argv[4]);

  if (mount(source, target, filesystemtype, mountflags, data) < 0)
    rb_sys_fail("mount");

  return INT2FIX(0);
}

/*
 * A wrapper around umount system call.
 *
 * == Synopsis:
 *   HrrRbMount.umount "target", flags # => 0
 *
 * @overload umount(target, flags=0) => 0
 *   @param target [String] The location (a directory or file) specified by the pathname.
 *   @param flags [Integer] The umount operation is performed depending on the bits specified in the flags.
 *   @return [Integer] 0.
 *   @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
 */
VALUE
hrr_rb_mount_umount(int argc, VALUE *argv, VALUE self)
{
  const char *target;
  int flags;

  rb_check_arity(argc, 1, 2);

  target = StringValueCStr(argv[0]);
  flags  = argc < 2 ? 0 : NUM2INT(argv[1]);

  if (umount2(target, flags) < 0)
    rb_sys_fail("umount");

  return INT2FIX(0);
}

void
Init_hrr_rb_mount(void)
{
  rb_mHrrRbMount = rb_define_module("HrrRbMount");

  rb_define_singleton_method(rb_mHrrRbMount, "mount",  hrr_rb_mount_mount,  -1);
  rb_define_singleton_method(rb_mHrrRbMount, "umount", hrr_rb_mount_umount, -1);

  rb_mHrrRbMountConst = rb_define_module_under(rb_mHrrRbMount, "Constants");
  rb_include_module(rb_mHrrRbMount, rb_mHrrRbMountConst);

  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "REMOUNT",     INT2NUM(MS_REMOUNT)    );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "BIND",        INT2NUM(MS_BIND)       );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "SHARED",      INT2NUM(MS_SHARED)     );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "PRIVATE",     INT2NUM(MS_PRIVATE)    );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "SLAVE",       INT2NUM(MS_SLAVE)      );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "UNBINDABLE",  INT2NUM(MS_UNBINDABLE) );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "MOVE",        INT2NUM(MS_MOVE)       );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "DIRSYNC",     INT2NUM(MS_DIRSYNC)    );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "LAZYTIME",    INT2NUM(MS_LAZYTIME)   );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "MANDLOCK",    INT2NUM(MS_MANDLOCK)   );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "NOATIME",     INT2NUM(MS_NOATIME)    );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "NODEV",       INT2NUM(MS_NODEV)      );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "NODIRATIME",  INT2NUM(MS_NODIRATIME) );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "NOEXEC",      INT2NUM(MS_NOEXEC)     );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "NOSUID",      INT2NUM(MS_NOSUID)     );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "RDONLY",      INT2NUM(MS_RDONLY)     );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "REC",         INT2NUM(MS_REC)        );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "RELATIME",    INT2NUM(MS_RELATIME)   );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "SILENT",      INT2NUM(MS_SILENT)     );
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "STRICTATIME", INT2NUM(MS_STRICTATIME));
  /* This flag can be specified in mount. */
  rb_define_const(rb_mHrrRbMountConst, "SYNCHRONOUS", INT2NUM(MS_SYNCHRONOUS));

  /* This flag can be specified in umount. */
  rb_define_const(rb_mHrrRbMountConst, "FORCE",    INT2NUM(MNT_FORCE)      );
  /* This flag can be specified in umount. */
  rb_define_const(rb_mHrrRbMountConst, "DETACH",   INT2NUM(MNT_DETACH)     );
  /* This flag can be specified in umount. */
  rb_define_const(rb_mHrrRbMountConst, "EXPIRE",   INT2NUM(MNT_EXPIRE)     );
  /* This flag can be specified in umount. */
  rb_define_const(rb_mHrrRbMountConst, "NOFOLLOW", INT2NUM(UMOUNT_NOFOLLOW));
}

File: lib/hrr_rb_mount.rb
require "hrr_rb_mount/version"
require "hrr_rb_mount/hrr_rb_mount"

# A wrapper around mount and umount. See mount(2) and umount(2) for details.
#
# @example
#   require "hrr_rb_mount"
#
#   flags = HrrRbMount::NOEXEC
#   HrrRbMount.mount "tmpfs", "/path/to/target", "tmpfs", flags, "size=1M" # => 0
#   HrrRbMount.mountpoint? "/path/to/target"                               # => true
#   HrrRbMount.umount "/path/to/target"                                    # => 0
#
module HrrRbMount

  # Constants that represent the flags for mount and umount operations.
  module Constants
  end

  # The path to /proc/self/mountinfo.
  PROC_MOUNTINFO_PATH = "/proc/self/mountinfo"

  # A wrapper around mount --remount mountpoint command.
  #
  # @example
  #   HrrRbMount.remount "mountpoint" # => 0
  #
  # @param mountpoint [String] The location (a directory or file) specified by the pathname.
  # @param mountflags [Integer] The umount operation is performed depending on the bits specified in the mountflags.
  # @param data [String] Per filesystem options.
  # @return [Integer] 0.
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.remount mountpoint, mountflags=0, data=""
    mount "none", mountpoint, nil, REMOUNT | mountflags, data
  end

  # A wrapper around mount --move source target command.
  #
  # @example
  #   HrrRbMount.move "source", "target" # => 0
  #
  # @param source [String] The pathname referring to a device or a pathname of a directory or file, or a dummy string.
  # @param target [String] The location (a directory or file) specified by the pathname.
  # @param mountflags [Integer] The umount operation is performed depending on the bits specified in the mountflags.
  # @param data [String] Per filesystem options.
  # @return [Integer] 0.
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.move source, target, mountflags=0, data=""
    mount source, target, nil, MOVE | mountflags, data
  end

  # A wrapper around mount --bind source target command.
  #
  # @example
  #   HrrRbMount.bind "source", "target" # => 0
  #
  # @param source [String] The pathname referring to a device or a pathname of a directory or file, or a dummy string.
  # @param target [String] The location (a directory or file) specified by the pathname.
  # @param mountflags [Integer] The mount operation is performed depending on the bits specified in the mountflags.
  # @param data [String] Per filesystem options.
  # @return [Integer] 0.
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.bind source, target, mountflags=0, data=""
    mount source, target, nil, BIND | mountflags, data
  end

  # A wrapper around mount --rbind source target command.
  #
  # @example
  #   HrrRbMount.rbind "source", "target" # => 0
  #
  # @param source [String] The pathname referring to a device or a pathname of a directory or file, or a dummy string.
  # @param target [String] The location (a directory or file) specified by the pathname.
  # @param mountflags [Integer] The mount operation is performed depending on the bits specified in the mountflags.
  # @param data [String] Per filesystem options.
  # @return [Integer] 0.
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.rbind source, target, mountflags=0, data=""
    mount source, target, nil, BIND | REC | mountflags, data
  end

  # A wrapper around mount --make-shared mountpoint command.
  #
  # @example
  #   HrrRbMount.make_shared "mountpoint" # => 0
  #
  # @param mountpoint [String] The location (a directory or file) specified by the pathname.
  # @return [Integer] 0.
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.make_shared mountpoint
    mount "none", mountpoint, nil, SHARED
  end

  # A wrapper around mount --make-slave mountpoint command.
  #
  # @example
  #   HrrRbMount.make_slave "mountpoint" # => 0
  #
  # @param mountpoint [String] The location (a directory or file) specified by the pathname.
  # @return [Integer] 0.
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.make_slave mountpoint
    mount "none", mountpoint, nil, SLAVE
  end

  # A wrapper around mount --make-private mountpoint command.
  #
  # @example
  #   HrrRbMount.make_private "mountpoint" # => 0
  #
  # @param mountpoint [String] The location (a directory or file) specified by the pathname.
  # @return [Integer] 0.
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.make_private mountpoint
    mount "none", mountpoint, nil, PRIVATE
  end

  # A wrapper around mount --make-unbindable mountpoint command.
  #
  # @example
  #   HrrRbMount.make_unbindable "mountpoint" # => 0
  #
  # @param mountpoint [String] The location (a directory or file) specified by the pathname.
  # @return [Integer] 0.
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.make_unbindable mountpoint
    mount "none", mountpoint, nil, UNBINDABLE
  end

  # A wrapper around mount --make-rshared mountpoint command.
  #
  # @example
  #   HrrRbMount.make_rshared "mountpoint" # => 0
  #
  # @param mountpoint [String] The location (a directory or file) specified by the pathname.
  # @return [Integer] 0.
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.make_rshared mountpoint
    mount "none", mountpoint, nil, REC | SHARED
  end

  # A wrapper around mount --make-rslave mountpoint command.
  #
  # @example
  #   HrrRbMount.make_rslave "mountpoint" # => 0
  #
  # @param mountpoint [String] The location (a directory or file) specified by the pathname.
  # @return [Integer] 0.
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.make_rslave mountpoint
    mount "none", mountpoint, nil, REC | SLAVE
  end

  # A wrapper around mount --make-rprivate mountpoint command.
  #
  # @example
  #   HrrRbMount.make_rprivate "mountpoint" # => 0
  #
  # @param mountpoint [String] The location (a directory or file) specified by the pathname.
  # @return [Integer] 0.
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.make_rprivate mountpoint
    mount "none", mountpoint, nil, REC | PRIVATE
  end

  # A wrapper around mount --make-runbindable mountpoint command.
  #
  # @example
  #   HrrRbMount.make_runbindable "mountpoint" # => 0
  #
  # @param mountpoint [String] The location (a directory or file) specified by the pathname.
  # @return [Integer] 0.
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.make_runbindable mountpoint
    mount "none", mountpoint, nil, REC | UNBINDABLE
  end

  # Returns true if the target directory or file is a mountpoint. Otherwise, returns false.
  #
  # Internally, uses /proc/self/mountinfo file to detect if the target is a mountpoint.
  #
  # When the file is not available, then uses stat(2).
  # In this case, bind mount is not able to be detected.
  #
  # @example
  #   HrrRbMount.mountpoint? "/proc"   # => true
  #   HrrRbMount.mountpoint? "/proc/1" # => false
  #
  # @param target [String] The target path to be checked.
  # @param follow_symlinks [Boolean] Specifies whether to follow symlinks.
  # @return [Boolean] true or false
  # @raise [Errno::EXXX] A SystemCallError is raised when the operation failed.
  def self.mountpoint? target, follow_symlinks: true
    return false if (! follow_symlinks) && File.symlink?(target)
    begin
      if File.exist? PROC_MOUNTINFO_PATH
        tgt_abs_path = File.realpath(target)
        File.foreach(PROC_MOUNTINFO_PATH){ |line|
          break true if line.split(" ")[4] == tgt_abs_path
        } or false
      else
        parent = File.join(target, "..")
        st, pst = File.stat(target), File.stat(parent)
        st.dev != pst.dev || st.ino == pst.ino
      end
    rescue Errno::ENOENT, Errno::ENOTDIR
      false
    end
  end
end
